### ISSUE TITLE
The chia harvester keeps consuming too much memory

### ISSUE THREAD

**Issue by ggvg-d** [madMAx43v3r/chia-gigahorse#177]
ubuntu23.04(6.2.0-25-generic)
8Gb RAM
Driver Version: 535.54.03    CUDA Version: 12.2 RTX3070
plots disk filesystem : NTFS (ntfs-3g mount readonly)
chia version（1.8.2）
chia-gigahorse-farmer-1.8.2.giga14-x86_64（chia farmer / harvester）
chia-gigahorse-farmer-1.8.1.giga13-x86_64（chia farmer）

Phenomenon：
1. Preparing a bunch of **uncompressed** NFT K32 plots drawn by Bladebit.
2. Stuffing these  plots into multiple NTFS hard drives.
3. As more and more uncompressed plots are loaded by Giga harvester, the more memory is clearly consumed, eventually stopping at a certain level.
4. Here are some examples I tested ：with 2167  plots, the memory eventually takes up around 5700MB，2609 plots 6615MB，and so on.(To be honest, this level of memory usage is hard to accept)
5. But for Chia's harvester, even with around 16,000 plots, it only needs about 1GB of memory in the end,(This includes the total memory usage of the desktop system)

**Comment by ggvg-d** [comment]
config.yaml use chia default

**Comment by Ealrann** [comment]
I have this problem too (well I suppose it's the same) : the memory consumption increases forever, until I reboot the server. Since I have 32GB, I don't reach OOM for now. 

20k plots, C7, ext4 largefile4, Archlinux, last version of madmax gigahorse farmer. 
I tried to tweek the number of thread, didn't solve the problem.

Here a monitorix of memory usage for the last month:
![image](https://github.com/madMAx43v3r/chia-gigahorse/assets/723000/82a82932-37ea-4092-9caf-fdddc942d549)

**Comment by madMAx43v3r** [comment]
If I remember correctly this is a known problem with Bladebit plots, because C2 table is too large and it's buffered in RAM.

**Comment by madMAx43v3r** [comment]
> 20k plots, C7, GPU farming (3060 ti), ext4 largefile4, Archlinux, 1.8.2.giga14

Is this a new problem with giga14?

**Comment by madMAx43v3r** [comment]
@ggvg-d

Please try this new chiapos.so, just replace yours in `chia-gigahorse-farmer` (after unzipping it): 
[chiapos.so.zip](https://github.com/madMAx43v3r/chia-gigahorse/files/12227309/chiapos.so.zip)

**Comment by Ealrann** [comment]
> > 20k plots, C7, GPU farming (3060 ti), ext4 largefile4, Archlinux, 1.8.2.giga14
> 
> Is this a new problem with giga14?

No, I'm seen this since I started with gigahorse on 1.6.x if I remember properly (in April)

**Comment by madMAx43v3r** [comment]
> No, I'm seen this since I started with gigahorse on 1.6.x if I remember properly (in April)

Hmm there must be something special about your setup, this is not normal.

**Comment by stefan-lange** [comment]
Have only a tiny farm with 8k C8, A2000, XFS, Debian 12 (container), 1.8.2.giga14 and it is pretty solid. The setup is not native, it is within kubernetes with memory limits. But it is stable since the beginning, after I found good limits for my case. Maybe it helps

![grafik](https://github.com/madMAx43v3r/chia-gigahorse/assets/1436672/0ce7c34b-3371-4196-9ddf-da1c2b6a26ec)

**Comment by Ealrann** [comment]
> > No, I'm seen this since I started with gigahorse on 1.6.x if I remember properly (in April)
> 
> Hmm there must be something special about your setup, this is not normal.

Yes, probably, but I don't find anything 😔 . The very high number of harvester processes makes it hard to debug for me
I suspect it's gigahorse because when I stop chia `chia stop all -d`, all the memory usage goes almost to zero.

![image](https://github.com/madMAx43v3r/chia-gigahorse/assets/723000/3f9699e9-33ff-4f0f-b9aa-ef2363bfcee1)

**Comment by madMAx43v3r** [comment]
Can you try flexfarmer? That would tell us if the problem is in my code or on Chia's python side.

**Comment by stefan-lange** [comment]
Do you know exactly that the harvester is the problem? The database is also a good 'memory eater'.

For a fast overview, the following script may help
https://github.com/shovon8/sysmon/blob/master/sysmon

Command `sudo ./sysmon | grep chia`

Example output
![grafik](https://github.com/madMAx43v3r/chia-gigahorse/assets/1436672/f7e96bc7-a082-421a-9e03-8d3fc14db544)

**Comment by Ealrann** [comment]
@madMAx43v3r 
I'm not on flex pool, but I'll give it a try if it can help :+1: 

@stefan-lange 
Thanks for the script ! 
Here the result : 
![image](https://github.com/madMAx43v3r/chia-gigahorse/assets/723000/2974be1b-d218-4583-b0b1-20a787fa775c)

I'll do it again in a few days to compare

**Comment by ggvg-d** [comment]
> @ggvg-d
> 
> Please try this new chiapos.so, just replace yours in `chia-gigahorse-farmer` (after unzipping it): [chiapos.so.zip](https://github.com/madMAx43v3r/chia-gigahorse/files/12227309/chiapos.so.zip)

After using this for a day, it looks like it solves the problem well. Now whether handling uncompressed plots or a mix of compressed and uncompressed, the harvester behaves normally. I will continue monitoring and provide an update if any new issues arise. Overall this is working much better, thanks for addressing this!

**Comment by ggvg-d** [comment]
The issue I was seeing seems to be resolved for now. If others run into other problems please open separate issues. I think we can likely close this issue @Ealrann . 

And thank you for the quick responses. @madMAx43v3r @stefan-lange

**Comment by madMAx43v3r** [comment]
Thanks for testing the fix, I will now include it in the next release.
